src/App/Fossa/PathDependency.hs

```diff
@@ -0,0 +1,258 @@
+
+      -- 3. We scan and upload dependencies, and retrieve transformed
+      -- dependencies!
+      scannedDeps <- traverse (scanAndUpload Nothing fullFile projectRevision) needToScanDepsWithMetadata
+
+      -- 4. Kickoff job to finalize the build process for path dependencies
+      let resolvedLoc = resolvedLocators scannedDeps
+      unless (null resolvedLoc) $
+        finalizeLicenseScanForPathDependency resolvedLoc False
+
+      -- 5. Finally update the graph with resolved path dependencies!
+      pure $ replaceResolved scannedDeps
+  where
+    replaceResolved :: [(Dependency, Maybe Dependency)] -> Graphing Dependency
+    replaceResolved deps = gmap (replace deps) graph
+
+    replace :: [(Dependency, Maybe Dependency)] -> Dependency -> Dependency
+    replace deps dep = case find (\d -> fst d == dep) deps of
+      Nothing -> dep
+      Just (_, maybeDep) -> fromMaybe dep maybeDep
+
+    resolvedLocators :: [(Dependency, Maybe Dependency)] -> [Locator]
+    resolvedLocators deps = map toLocator (mapMaybe snd deps)
+
+-- | Performs license scanning and upload for single path dependency.
+scanAndUpload ::
+  ( Has Diagnostics sig m
+  , Has (Lift IO) sig m
+  , Has Exec sig m
+  , Has ReadFS sig m
+  , Has StickyLogger sig m
+  , Has FossaApiClient sig m
+  ) =>
+  Maybe LicenseScanPathFilters ->
+  FullFileUploads ->
+  ProjectRevision ->
+  (Dependency, SomeResolvedPath, Text) ->
+  m (Dependency, Maybe Dependency)
+scanAndUpload pathFilters fullFileUpload projectRevision (rawDep, resolvedPath, version) = context "Path Dependency" $ do
+  maybeLicenseUnits <- recover scan
+  case maybeLicenseUnits of
+    Nothing -> pure (rawDep, Nothing)
+    Just licenseUnits -> (rawDep,) <$> upload licenseUnits
+  where
+    rawPath :: Text
+    rawPath = dependencyName rawDep
+
+    scan = do
+      logSticky $ "License Scanning '" <> rawPath <> "' at '" <> toText resolvedPath <> "'"
+      licenseUnits <- case resolvedPath of
+        ResolvedDir dir -> scanDirectory Nothing mempty pathFilters fullFileUpload dir
+        ResolvedFile _ -> fatalText $ "path dependency for single file is not supported: " <> rawPath
+      pure $ LicenseSourceUnit rawPath CliLicenseScanned licenseUnits
+
+    upload licSrcUnit = do
+      logSticky $ "Uploading '" <> rawPath <> "' to secure S3 bucket"
+      resp <- uploadPathDependencyScan (PackageRevision rawPath version) projectRevision fullFileUpload
+
+      let signedURL = pdSignedURL resp
+      let name' = updlName . pdLocator $ resp
+      let version' = updlVersion . pdLocator $ resp
+      uploadLicenseScanResult signedURL licSrcUnit
+
+      pure $
+        Just
+          rawDep
+            { dependencyType = PathType
+            , dependencyName = name'
+            , dependencyVersion = Just $ CEq version'
+            }
+
+-- * Helpers
+
+canResolve :: Dependency -> Bool
+canResolve d = dependencyType d == UnresolvedPathType
+
+isValidDep :: Bool -> Dependency -> Bool
```

Comment:
Can we rename this to `isValidPathDep` and `canResolve` to `canResolvePathDep` or something like that? I think this would reduce uncertainty that these functions are only for path deps.